powerauth-restful-security-javaee/src/main/java/io/getlime/security/powerauth/rest/api/jaxrs/controller/v3/TokenController.java

```diff
@@ -0,0 +1,182 @@
+                final String mac = requestObject.getMac();
+
+                // Prepare a signature type converter
+                SignatureTypeConverter converter = new SignatureTypeConverter();
+
+                // Get ECIES headers
+                String applicationKey = encryption.getApplicationKey();
+                String activationId = encryption.getActivationId();
+
+                // Create a token
+                final PowerAuthPortV3ServiceStub.CreateTokenResponse token = powerAuthClient.createToken(activationId, applicationKey, ephemeralPublicKey,
+                        encryptedData, mac, converter.convertFrom(signatureFactors));
+
+                // Prepare a response
+                final io.getlime.security.powerauth.rest.api.model.response.v3.TokenCreateResponse responseObject = new io.getlime.security.powerauth.rest.api.model.response.v3.TokenCreateResponse();
+                responseObject.setMac(token.getMac());
+                responseObject.setEncryptedData(token.getEncryptedData());
+                return new ObjectResponse<>(responseObject);
+            } else {
+                throw new PowerAuthAuthenticationException();
+            }
+        }  catch (PowerAuthAuthenticationException ex) {
+            throw ex;
+        } catch (Exception ex) {
+            throw new PowerAuthAuthenticationException(ex.getMessage());
+        }
+
+    }
+
+    @POST
+    @Consumes({MediaType.APPLICATION_JSON})
+    @Produces({MediaType.APPLICATION_JSON})
+    @Path("remove")
+    public ObjectResponse<TokenRemoveResponse> removeToken(ObjectRequest<TokenRemoveRequest> request, @HeaderParam(PowerAuthTokenHttpHeader.HEADER_NAME) String tokenHeader) throws PowerAuthAuthenticationException {
+
+        try {
+            PowerAuthApiAuthentication authentication = authenticationProvider.validateToken(tokenHeader, Arrays.asList(
+                    PowerAuthSignatureTypes.POSSESSION,
+                    PowerAuthSignatureTypes.POSSESSION_KNOWLEDGE,
+                    PowerAuthSignatureTypes.POSSESSION_BIOMETRY,
+                    PowerAuthSignatureTypes.POSSESSION_KNOWLEDGE_BIOMETRY
+            ));
+
+            if (authentication != null && authentication.getActivationId() != null) {
+
+                // Fetch activation ID
+                final String activationId = authentication.getActivationId();
+
+                // Fetch token ID from the request
+                final TokenRemoveRequest requestObject = request.getRequestObject();
+                final String tokenId = requestObject.getTokenId();
+
+                // Remove a token, ignore response, since the endpoint should quietly return
+                powerAuthClient.removeToken(tokenId, activationId);
+
+                // Prepare a response
+                final TokenRemoveResponse responseObject = new TokenRemoveResponse();
+                responseObject.setTokenId(requestObject.getTokenId());
```

Comment:
... you already have `tokenId` variable couple rows above...

Reply:
It's taken from version 2.0 `removeToken` implementation, I will fix it in all places.